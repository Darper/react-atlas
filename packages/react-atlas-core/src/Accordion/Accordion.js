import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";

/**
 * Avatar component creates a circular area where an image, letter or icon/glyphicon can be presented. Great for user profiles and lists.
 *
 * **NOTE**: children will always take precedence over props passed into component.
 */
class Accordion extends React.PureComponent {
    constructor(props) {
        super(props);

        // array used to set the activeChildArray value for each child for expanding purposes
        let childArray = [];

        // loop through children and set childArray depending on if expanded prop was set on child.  If not set this defaults to false
        let childCount = React.Children.map(this.props.children, (child, i) => {
            if(child.props.expanded){
                childArray.push(true);
            } else {
                childArray.push(false);
            }
        })

        // component state
        this.state = {
            title: props.title,
            activeChildArray: childArray,
            expandAll: this.props.expandAll || false,
            expanded: false,
            multiOpen: this.props.multiOpen || false
        };
    }

    // click handler for header.  Sets activeChildArray value accordingly to expand and collapse the panels
    _click = (idx) => {
        let newChildArray = [];
        for(var i = 0; i < this.state.activeChildArray.length; i++) {
            if(i === idx) {
                newChildArray.push(!this.state.activeChildArray[i])
            } else if (!this.state.multiOpen){
                newChildArray.push(false)
            } else {
                newChildArray.push(this.state.activeChildArray[i])
            }
        }
        this.setState({
            activeChildArray: newChildArray,
        });
    };

    // expandAll function to expand all panels.  Sets value of all to true in activeChildArray
    _expandAll = () => {
        let newChildArray = [];
        for(var i = 0; i < this.state.activeChildArray.length; i++) {
            newChildArray.push(!this.state.expanded);
        }
        this.setState({
            activeChildArray: newChildArray,
            expanded: !this.state.expanded
        });
    }

    render() {
        let { className, children, title, accordionWidth } = this.props;

        // activeClass is added to activ
        let activeClass;

        // A list of children of the Accordion component
        const accordion_panels = React.Children.map(children, (child, i) => {

            if(this.state.activeChildArray[i] === true){
                activeClass = cx({
                    ra_accordion__active: true
                });
            } else {
                activeClass = "";
            }

            let accordion_panel = (
                <div>
                    <div
                        styleName={"accordion_header"}
                        className={activeClass}
                        onClick={() => {
                            this._click(i);
                        }}>
                            {child.props.title}
                    </div>
                    <div styleName={"accordion_panel"} className={activeClass}>{child}</div>
                </div>
            );
            return accordion_panel;
        });

        return (
            <div>
                { this.state.expandAll ? <div styleName={"expandAll"} onClick={() => {this._expandAll()}}>Expand All</div> : null }
                <div styleName={"accordion"} style={{ width: accordionWidth + "px" }}>
                    {accordion_panels}
                </div>
            </div>
        );
    }
}

Accordion.propTypes = {
    /**
     * Children should be either a string, an icon/glyphicon, or an image tag.
     * @examples "SomeName", <SomeIcon />, <img src="/path/to/image.jpg"/>
     */
    children: PropTypes.node,
    /**
     *
     */
    className: PropTypes.string,
    /**
     * A string. Accordion will use First letter of the string.
     * @examples "Nathan" will output "N"
     */
    title: PropTypes.string
};

Accordion.defaultProps = {
    accordionWidth: "300",
    disabled: false
};

export default Accordion;