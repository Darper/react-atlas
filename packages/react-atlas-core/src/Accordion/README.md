Title prop gets truncated to 1st letter:

     <Accordion>
          <div title="first">this is the first panelasdasdasdas asdasd asda asda asd asd asdas dasda asdddd sad sd a asdadasd asd ad dsadasd asd asdasd ad asd </div>
          <div title="second">this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>


Accordion with expand all

     <Accordion expandAll={true}>
          <div title="first">this is the first panel</div>
          <div title="second">this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>

Accordion with multiOpen functionality

     <Accordion multiOpen={true}>
          <div title="first">this is the first panel</div>
          <div title="second">this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>

Accordion with Width set in props

     <Accordion accordionWidth={400}>
          <div title="first">this is the first panel</div>
          <div title="second">this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>

Accordion with defined open panel

     <Accordion>
          <div title="first">this is the first panel</div>
          <div title="second" expanded>this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>

Accordion with disabled prop

     <Accordion disabled>
          <div title="first">this is the first panel</div>
          <div title="second">this is the second panel</div>
          <div title="third">this is the third panel</div>
     </Accordion>