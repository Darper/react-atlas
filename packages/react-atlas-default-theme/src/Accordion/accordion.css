.accordion {
    composes: default-border from '../styles.css';
    composes: border-bottom-none from '../styles.css';
}
.accordion_header {
    composes: marg-0  from '../styles.css';
    composes: default-text from '../styles.css';
    composes: bold from '../styles.css';
    composes: cursor-pointer from '../styles.css';
    composes: bg-silver from '../styles.css';
    composes: default-border-bottom from '../styles.css';
    composes: border-top-none from '../styles.css';
    composes: border-left-none from '../styles.css';
    composes: border-right-none from '../styles.css';
    line-height: 1.5rem;
    vertical-align: middle;
    text-align: center;
}
.accordion_header.active {
    border-bottom: 0;
    background-color: #00a7e1;
    color: white;
}
.accordion_header:hover {
    background-color: #ccc;
    color: white;
}.accordion_header.active:hover {
     background-color: #00a7e1;
 }

.accordion_panel {
    composes: default-text from '../styles.css';
    composes: border-top-none from '../styles.css';
    composes: border-bottom-none from '../styles.css';
    composes: border-left-none from '../styles.css';
    composes: border-right-none from '../styles.css';
    max-height: 0;
}
.accordion_panel div {
    display: none;
}
.accordion_panel.active{
    border-top: none;
    border-left: none;
    border-right: none;
    border-bottom: 1px solid #999;
    max-height: 100%;
    transition: max-height .25s ease-in-out;
    -moz-transition: max-height .25s ease-in-out;
    -webkit-transition: max-height .25s ease-in-out;
    padding: 1rem;
}
.accordion_panel.active div {
    display: block;
}

.expandAll {
    composes: marg-b-1 from '../styles.css';
    composes: default-text from '../styles.css';
    composes: primary from '../styles.css';
    cursor: pointer;
}
.expandAll:hover {
    text-decoration: underline;
}