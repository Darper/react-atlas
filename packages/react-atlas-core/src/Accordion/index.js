import { default as AccordionCore } from "./Accordion";
export { AccordionCore };
