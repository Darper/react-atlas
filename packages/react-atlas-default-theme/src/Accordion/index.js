import * as AccordionStyle from "./accordion.css";
export { AccordionStyle };
