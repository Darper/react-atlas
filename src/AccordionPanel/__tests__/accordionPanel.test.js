import React from "react";
import { mount } from "enzyme";
import { AccordionPanel } from "../index";
import { Accordion } from "../../Accordion/index";
import renderer from "react-test-renderer";

describe("Test Accordion Panel component", () => {
  it("Test render correctly", () => {
    const comp = 
      <AccordionPanel title="First">
        Some content for the first panel
      </AccordionPanel>
    ;
    const tree = renderer.create(comp).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders children", () => {
    const ap = mount(
      <AccordionPanel>
        <span>Some content for the first panel</span>
        <span>Some more content for the first panel</span>
      </AccordionPanel>
    );

    expect(ap.props().children.length).toBe(2);
    expect(ap.props().children[0].type).toBe("span");
    expect(ap.props().children[0].props.children).toBe(
      "Some content for the first panel"
    );
    expect(ap.props().children[1].type).toBe("span");
    expect(ap.props().children[1].props.children).toBe(
      "Some more content for the first panel"
    );
    expect(ap.find("span").length).toBe(2);
  });

  it("renders className", () => {
    const ap = mount(
      <AccordionPanel className="foo">
        <span>Some content for the first panel</span>
        <span>Some more content for the first panel</span>
      </AccordionPanel>
    );

    expect(ap.find(".foo").exists()).toBe(true);
  });

  it("handles expanded", () => {
    const ap = mount(
      <Accordion>
        <AccordionPanel expanded>
          <span>Some content for the first panel</span>
          <span>Some more content for the first panel</span>
        </AccordionPanel>
      </Accordion>
    );

    expect(ap.find(".active").exists()).toBe(true);
  });

  it("is not expanded by default", () => {
    const ap = mount(
      <Accordion>
        <AccordionPanel>
          <span>Some content for the first panel</span>
          <span>Some more content for the first panel</span>
        </AccordionPanel>
      </Accordion>
    );

    expect(ap.find(".active").exists()).toBe(false);
    expect(ap.find(".inactive").exists()).toBe(true);
    expect(ap.find(".header-inactive").exists()).toBe(true);
  });

  it("renders style", () => {
    const ap = mount(
      <AccordionPanel style={{ "color": "red" }}>
        <span>Some content for the first panel</span>
        <span>Some more content for the first panel</span>
      </AccordionPanel>
    );

    expect(ap.props().style).toEqual({ "color": "red" });
  });

  it("renders title", () => {
    const ap = mount(
      <Accordion>
        <AccordionPanel title="First Panel">
          <div>Some content for the first panel</div>
          <div>Some more content for the first panel</div>
        </AccordionPanel>
      </Accordion>
    );
    expect(ap.instance().props.children.props.title).toBe("First Panel");
    expect(ap.find(".accordion_header").props().children).toBe("First Panel");
  });
});
