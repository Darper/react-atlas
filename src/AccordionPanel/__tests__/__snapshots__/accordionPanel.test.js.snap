// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Accordion Panel component Test render correctly 1`] = `
<div
  className="accordionPanel"
  style={undefined}
>
  Some content for the first panel
</div>
`;
